pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -423,102 +427,117 @@ func (a Adapter) waitAndGetComponentPod(hideSpinner bool) (*corev1.Pod, error) {
+						// Start the supervisord if the odo component does not exist
+						if !componentExists {
+							err = a.InitRunContainerSupervisord(*action.Component, podName, containers)
+							if err != nil {
+								return
+							}
+						}
+
+						// Exec the supervisord ctl stop and start for the devrun program
+						type devRunExecutable struct {
+							command []string
+						}
+						devRunExecs := []devRunExecutable{
+							{
+								command: []string{common.SupervisordBinaryPath, "ctl", "stop", "all"},
+							},
+							{
+								command: []string{common.SupervisordBinaryPath, "ctl", "start", string(common.DefaultDevfileRunCommand)},
+							},
+						}
+
+						s = log.Spinner("Executing " + command.Name + " command " + fmt.Sprintf("%q", *action.Command))
+						defer s.End(false)
+
+						for _, devRunExec := range devRunExecs {
+
+							err = exec.ExecuteCommand(&a.Client, podName, *action.Component, devRunExec.command, show)
+							if err != nil {
+								return
+							}
+						}
+						s.End(true)
+					}
 				}
-				s.End(true)
 			}
+		}
+	}
 
-			// Reset the for loop counter and iterate through all the devfile commands again for others
-			i = -1
-			// Set the buildRequired to false since we already executed the build command
-			buildRequired = false
-		} else if (command.Name == string(common.DefaultDevfileRunCommand) || command.Name == a.devfileRunCmd) && !buildRequired {
-			// Always check for buildRequired is false, since the command may be iterated out of order and we always want to execute devBuild first if buildRequired is true. If buildRequired is false, then we don't need to build and we can execute the devRun command
-			glog.V(3).Infof("Executing devfile command %v", command.Name)
-
-			for _, action := range command.Actions {
+	return
+}
 
-				// Check if the devfile run component containers have supervisord as the entrypoint.
-				// Start the supervisord if the odo component does not exist
-				if !componentExists {
-					err = a.InitRunContainerSupervisord(*action.Component, podName, containers)
-					if err != nil {
-						return
-					}
-				}
+func (a Adapter) executeDevfileCommand(command versionsCommon.DevfileCommand, show bool, podName string) error {
+	var s *log.Status
+	glog.V(3).Infof("Executing devfile command %v", command.Name)
 
-				// Exec the supervisord ctl stop and start for the devrun program
-				type devRunExecutable struct {
-					command []string
-				}
-				devRunExecs := []devRunExecutable{
-					{
-						command: []string{common.SupervisordBinaryPath, "ctl", "stop", "all"},
-					},
-					{
-						command: []string{common.SupervisordBinaryPath, "ctl", "start", string(common.DefaultDevfileRunCommand)},
-					},
-				}
+	for _, action := range command.Actions {
+		// Change to the workdir and execute the command
+		var cmdArr []string
+		if action.Workdir != nil {
+			cmdArr = []string{"/bin/sh", "-c", "cd " + *action.Workdir + " && " + *action.Command}
```

Comment:
`/bin/sh` should be a const.. See the code you deleted and how we used `common.SupervisordBinaryPath`